### PR TITLE
[XPU] Support checkpoint-engine weight update on Intel XPU

### PR DESCRIPTION
**Pull request by siju-samuel**
## Motivation
This PR adds Intel XPU as a first-class device backend behind device abstraction, and fills the IPC gap with a native SYCL ipc_memory transport, which is the XPU counterpart of CUDA IPC
The CUDA and NPU paths are unchanged, XPU-specific code is reached only when the device type is xpu.

This unblocks checkpoint-engine's fast weight-update path on Intel GPUs, which is needed for SGLang's XPU for RL rollout weight sync.

## Modifications
**Device abstraction**
- `device_utils.py`: detect `xpu` and map it to the `xccl` collective backend; add `supports_device_ipc()` / `supports_device_p2p()` capability gates so XPU enables broadcast IPC but not P2P.
- `distributed/base.py`: reject custom distributed backends on XPU and fall back to the default `xccl` TorchBackend.

**Weight transport **
- `transport.py`: add a `WeightTransport` seam — `IpcWeightTransport` (CUDA/NPU, unchanged) and `XpuIpcWeightTransport` (XPU) — with `build_transport()` selecting by device type.
- `xpu_ipc/sycl_ipc.cpp` + `xpu_ipc/__init__.py`: native SYCL `ipc_memory` extension (get/open/close), JIT-compiled at runtime with `-fsycl`, providing the cross-process device-buffer IPC that PyTorch lacks on XPU.

**Parameter server & worker**
- `ps.py`: route the broadcast through `build_transport()`, prebuild the SYCL extension at startup, and fail loudly with clear errors when device IPC/P2P is unavailable on XPU.
- `worker.py`: derive the `GPU-<uuid>` handshake key from `torch.xpu` on XPU, matching the ParameterServer's `_get_physical_gpu_id`.

**Packaging**
- `pyproject.toml`: ship the SYCL `.cpp` source as package data (JIT-compiled on XPU hosts; inert on CUDA).

**Tests**
- CPU-only: `test_device_manager.py`, `test_transport.py`, `test_xpu_parity.py`, `test_p2p_guard.py` — device dispatch, transport seam, portable-handle contract, `xccl` selection, P2P rejection.
- Hardware-gated: `test_xpu_ipc.py` — SYCL IPC roundtrip, interior-offset, full cross-process broadcast.

## Accuracy Test / Verified
Device-UUID handshake verified on Intel Arc Pro B60: the worker emits the exact
`GPU-<uuid>` key the ParameterServer expects. 5 CPU unit tests pass.

## Environment

| Software | Version |
| :--- | :--- |
| torch | 2.13.0.dev+xpu |
| oneAPI (icpx) | 2026.0 / 2026.1 |
| device | 8× Intel Arc Pro B60 |

## Check list
- [x] Code self-tested on Intel XPU hardware
- [x] CPU unit tests: 38 passed (`pytest tests/ -m "not gpu"`)
- [x] Hardware tests: 3 passed incl. cross-process broadcast (`pytest tests/test_xpu_ipc.py`)
- [x] CUDA/NPU paths unchanged (XPU code reached only when device type is `xpu`)

CC: @@kip-cxj @specture724 @weixiao-huang  @ppwwyyxxc Could you please run the CI and review. TIA

### CODE REVIEW

File: checkpoint_engine/xpu_ipc/__init__.py [hunk 1]
```diff
@@ -0,0 +1,163 @@
+"""Cross-process device-buffer IPC for Intel XPU via SYCL ``ipc_memory``.
+
+``sycl_ipc.cpp`` wraps ``ipc_memory`` (``get``/``open``/``close``), exported by
+torch's own libsycl (oneAPI >= 2026.0); this module JIT-compiles it with
+``-fsycl``. The handle is a self-contained portable byte blob (no dma-buf fd, no
+offset to carry), so it rides the existing ZMQ channel like CUDA's
+``reduce_tensor`` tuple -- see ``XpuIpcWeightTransport``.
+"""
+
+from __future__ import annotations
```
Comment (weixiao-huang) [inline checkpoint_engine/xpu_ipc/__init__.py]: It is unnecessary to import this line. I thinks we can use double quote in type

**Comment by siju-samuel** [comment]
@kip-cxj @specture724 @weixiao-huang  @ppwwyyxx 
Could you share your views on this PR.

File: checkpoint_engine/device_utils.py [hunk 2]
```diff
@@ -255,3 +271,38 @@ def rdma_device(self, rank: int) -> str:
             return _get_my_rdma_device(rank, self.device_module.device_count(), _get_rdma_devices())
         else:
             raise TypeError("The current transfer engine protocol is not supported")
+
+    def ipc_collect(self) -> None:
+        """Reclaim memory held by stale IPC handles where the backend supports it (no-op otherwise)."""
+        fn = getattr(self.device_module, "ipc_collect", None)
+        if callable(fn):
```
Comment (weixiao-huang) [inline checkpoint_engine/device_utils.py]: If not callable, mayeb it should raise

File: checkpoint_engine/ps.py [hunk 3]
```diff
@@ -780,112 +826,123 @@ def _update_per_bucket(
...
-                f"[rank{self._rank}] post-release: device mem usage: {(device_mem[1] - device_mem[0]) / 1024 / 1024:.2f} MB, "
-                f"allocated: {self.device_manager.device_module.memory_allocated() / 1024 / 1024:.2f} MB, "
-                f"reserved: {self.device_manager.device_module.memory_reserved() / 1024 / 1024:.2f} MB"
-            )
-            # Notify worker to call post_hook
-            socket.send_pyobj(None)
-            socket.recv()
+            gidx = 0
+            ret_code = torch.zeros((), device=self.device_manager.device_type, dtype=torch.int64)
+            try:
```
Comment (weixiao-huang) [inline checkpoint_engine/ps.py]: Why there need additinal `try` and `finally`?

File: checkpoint_engine/ipc_handler.py [hunk 4]
```diff
@@ -0,0 +1,124 @@
+"""Pluggable device-buffer handoff between the ParameterServer and the worker.
```
Comment (HubertZhang) [inline checkpoint_engine/ipc_handler.py]: What about renaming  WeightTransport to IPCHandler? This object transports nothing; it just hands an IPC handle for a device buffer across processes.

File: checkpoint_engine/xpu_ipc/sycl_ipc.cpp [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+#include <sycl/ext/oneapi/experimental/ipc_memory.hpp>
+
+#include <cstddef>
+#include <cstdint>
+#include <mutex>
+#include <optional>
+#include <unordered_map>
+#include <vector>
+
+namespace ipc = sycl::ext::oneapi::experimental::ipc_memory;
```
Comment (HubertZhang) [inline checkpoint_engine/xpu_ipc/sycl_ipc.cpp:25]: It seems `ipc_memory` namespace is deprecated upstream and should migrate to `ipc::memory`?

File: checkpoint_engine/xpu_ipc/__init__.py [hunk 6]
```diff
@@ -0,0 +1,163 @@
...
+    os.environ["CC"], os.environ["CXX"] = icx, icpx
+    try:
+        # Do NOT pin -std: torch.utils.cpp_extension injects the standard its ATen
+        # headers require, and a pin here would override it.
+        module = load(
+            name="checkpoint_engine_sycl_ipc",
+            sources=[str(src)],
+            extra_cflags=["-fsycl", "-O2", *sycl_include_flags],
+            extra_ldflags=["-fsycl"],
+            verbose=False,
```
Comment (HubertZhang) [inline checkpoint_engine/xpu_ipc/__init__.py:101]: Would this file be simplified if we set `with_sycl=True` here?

File: checkpoint_engine/xpu_ipc/sycl_ipc.cpp (hunk 5 above)
Comment (siju-samuel) [reply checkpoint_engine/xpu_ipc/sycl_ipc.cpp:25]: i have verified with oneapi 2026.1 version and it still doesnt have ipc::memory namespace

**Comment by siju-samuel** [comment]
@HubertZhang @weixiao-huang  Thanks a lot for the initial review. 
Could you please take another look. I have fixed and pushed the reworked code.

**Comment by HubertZhang** [comment]
LTGM

File: checkpoint_engine/xpu_ipc/__init__.py [hunk 7]
```diff
@@ -0,0 +1,145 @@
...
+
+    Compilers without the header are skipped: they build a device image that
+    torch's newer libsycl cannot load, aborting the process on dlopen rather
+    than raising something we could catch.
+    """
+    candidates: list[str] = []
+    root = os.getenv("CMPLR_ROOT")
+    if root:
+        candidates.append(os.path.join(root, "bin", "icpx"))
+    candidates += sorted(glob.glob("/opt/intel/oneapi/compiler/*/bin/icpx"), reverse=True)
```
Comment (weixiao-huang) [inline checkpoint_engine/xpu_ipc/__init__.py]: This sorted prefer the newest compiler. But current code may cause 2026.9 beats 2026.10.

Maybe you can use 
```python
def _icpx_version_key(path: str) -> list[int]:
    """Numeric version parts of ``.../compiler/<ver>/bin/icpx``, for newest-first sorting."""
    version = path.split("/")[-3]
    return [int(p) if p.isdigit() else -1 for p in version.split(".")]
```

and
```python
    candidates += sorted(
        glob.glob("/opt/intel/oneapi/compiler/*/bin/icpx"),
        key=_icpx_version_key,
        reverse=True,
    )
```

File: checkpoint_engine/ps.py [hunk 8]
```diff
@@ -780,112 +826,118 @@ def _update_per_bucket(
...
-        socket, socket_paths = self._bind_zmq_socket()
-        req_thread = threading.Thread(
-            target=req_func,
-            args=(socket_paths,),
-        )
-        req_thread.start()
-        socket.send_pyobj(handle)
+        # `with` releases the exported handle even if the bind or first send below
+        # fails; the barrier stays inside so an early failure cannot deadlock peers.
+        with build_ipc_handler(self.device_manager) as ipc_handler:
```
Comment (weixiao-huang) [inline checkpoint_engine/ps.py]: This `with` block takes too many changes to review. Another way may be wrapping this `with build_ipc_handler(self.device_manager) as ipc_handler` on `self._update_per_bucket` like
```python
            with build_ipc_handler(self.device_manager) as ipc_handle:
                self._update_per_bucket(checkpoint_name, req_func, ranks_group, ranks, ipc_handle)
```

**Review by weixiao-huang — COMMENTED** [review]
plz resolve the conflict from main branch

File: checkpoint_engine/xpu_ipc/__init__.py (hunk 7 above)
Comment (siju-samuel) [reply checkpoint_engine/xpu_ipc/__init__.py]: @weixiao-huang Thanks for finding it out. i have modifed this bug. 
Info: Usually oneapi will release only 4 versions in an year. 0-3 only can happen.

File: checkpoint_engine/ps.py (hunk 8 above)
Comment (siju-samuel) [reply checkpoint_engine/ps.py]: Done, This is much better than what I had. Thanks

File: checkpoint_engine/ps.py [hunk 9]
```diff
@@ -726,16 +754,38 @@ def _update_per_bucket(
         req_func: Callable[[list[tuple[str, str]]], None],
         ranks_group: dist.DistributedProcessGroup | None,
         ranks: list[int] | None = None,
+        ipc_handler: IPCHandler | None = None,
```
Comment (weixiao-huang) [inline checkpoint_engine/ps.py]: I think this should be set as a required args since you assert it's not None below. So maybe we should place this right after `req_func` and change signature to `ipc_handler: IPCHandler` and the test for `_update_per_bucket` should change also.
